pipelines/sql_database/util.py

```diff
@@ -0,0 +1,116 @@
+from typing import cast, TypedDict, Any, List, Optional, Mapping, Iterator, Dict
+
+from sqlalchemy import Table, tuple_
+from sqlalchemy.engine import Result, Engine
+from sqlalchemy.sql import Select
+
+import dlt
+
+
+class CursorState(TypedDict):
+    last_value: Any
+    loaded_ids: List[Any]
+
+
+
+class Cursor:
+    def __init__(self, table_name: str, cursor_column: str, unique_column: Optional[str]) -> None:
+        self.table_name = table_name
+        self.cursor_column = cursor_column
+        self.unique_column = unique_column
+
+    def get_state(self) -> CursorState:
+        states = dlt.state().setdefault('incremental', {})
+        table_state = states.setdefault(self.table_name, {})
+        cursor_state = table_state.setdefault(self.cursor_column, {'last_value': None, 'loaded_ids': []})
+        return cast(CursorState, cursor_state)
+
+    def update(self, cursor_value: Any, loaded_id: Any) -> None:
+        state = self.get_state()
+        current_val = state['last_value']
+        if current_val != cursor_value:
+            state['last_value'] = cursor_value
+            state['loaded_ids'] = []
+        if loaded_id is not None:
+            state['loaded_ids'].append(loaded_id)
+
+
+class TableLoader:
+    def __init__(
+        self,
+        engine: Engine,
+        table: Table,
+        cursor_column: Optional[str] = None,
+        unique_column: Optional[str] = None
+    ) -> None:
+        self.engine = engine
+        self.table = table
+        self.cursor = (
+            Cursor(table.name, cursor_column, unique_column)
+            if cursor_column is not None else None
+        )
+        self.cursor_column = table.c.get(cursor_column)
+        self.unique_column = table.c.get(unique_column)
+        self.table_has_cursor = self.cursor_column is not None
+        self.table_has_unique = self.unique_column is not None
+
+    def make_query(self) -> Select[Any]:
+        table = self.table
+        query = table.select()
+        if not self.cursor:
+            return query
+        cursor_col = self.cursor_column
+        if cursor_col is None:
```

Comment:
I'd raise exception if cursor/primary key do not exist

Reply:
Added this in `Cursor.__init__` so it's raised early as possible